kernel/src/process/pid_namespace/mod.rs

```diff
@@ -0,0 +1,303 @@
+        let id = PID_NS_ID_ALLOCATOR.fetch_add(1, Ordering::Relaxed);
+        Self {
+            id,
+            nested_level: 0,
+            pid_allocator: AtomicU32::new(1),
+            parent: Weak::new(),
+            children: Mutex::new(BTreeMap::new()),
+            pid_entry_map: Mutex::new(BTreeMap::new()),
+            is_init_proc_terminated: AtomicBool::new(false),
+        }
+    }
+
+    /// Creates a new child namespace.
+    pub fn new_child(parent: &Arc<Self>) -> Result<Arc<Self>> {
+        let nested_level = parent.nested_level + 1;
+        if nested_level >= MAX_NESTED_LEVEL {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "the namespace nested level has reached its limit"
+            );
+        }
+
+        let id = PID_NS_ID_ALLOCATOR.fetch_add(1, Ordering::Relaxed);
+        let child = Arc::new(Self {
+            id,
+            nested_level,
+            pid_allocator: AtomicU32::new(1),
+            parent: Arc::downgrade(parent),
+            children: Mutex::new(BTreeMap::new()),
+            pid_entry_map: Mutex::new(BTreeMap::new()),
+            is_init_proc_terminated: AtomicBool::new(false),
+        });
+        parent.children.lock().insert(id, child.clone());
+        Ok(child)
+    }
+
+    /// Allocates a [`AncestorNsPids`] from the namespace and all ancestor PID namespaces.
+    pub fn allocate_ids(self: &Arc<Self>) -> AncestorNsPids {
+        let _global_guard = PID_ENTRY_MAP_LOCK.write();
+
+        let mut namespace = self.clone();
+        let mut ids = VecDeque::new();
+
+        loop {
+            let id = namespace.pid_allocator.fetch_add(1, Ordering::Relaxed);
+            ids.push_front(NsPid {
+                id,
+                pid_ns: Arc::downgrade(&namespace),
+            });
+
+            if namespace.nested_level == 0 {
+                break;
+            }
+
+            namespace = namespace.parent.upgrade().unwrap();
+        }
+
+        let ids = AncestorNsPids(Arc::new(ids));
+        let pid_entry_inner = Arc::new(PidEntryInner::new_empty(ids.clone()));
+
+        namespace = self.clone();
+
+        loop {
+            let id = namespace.get_current_id(&ids).unwrap();
+            namespace
+                .pid_entry_map
+                .lock()
+                .insert(id, pid_entry_inner.clone());
+
+            if namespace.nested_level == 0 {
+                break;
+            }
+
+            namespace = namespace.parent.upgrade().unwrap();
+        }
+
+        ids
+    }
+
+    /// Gets the `TaskId` of the `ids` in this PID namespace.
+    ///
+    /// It the `ids` is not visible in `self`, this method will return `None`.
+    pub fn get_current_id(self: &Arc<Self>, ids: &AncestorNsPids) -> Option<TaskId> {
+        let ns_pid = ids.0.get(self.nested_level)?;
+
+        Weak::ptr_eq(&ns_pid.pid_ns, &Arc::downgrade(self)).then_some(ns_pid.id)
+    }
+
+    pub fn get_thread(&self, id: Tid) -> Option<Arc<Thread>> {
```

Comment:
Based on the previous comments, I believe that all of the `get_xxx() -> Option<Arc<Xxx>>` functions are susceptible to the same race conditions. I'm not sure if you can fix it in this PR since there are many uses, but I'm leaving this comment for the record.